Update gold standard related functions

hagenw opened the pull request:
Closes #7

This updates the input and output types of ratings to the functions. We now support array like data as input and return float or an array instead of dataframes. It further splits the gold standard and confidence functions into separate functions and removes `mean()` and `median()` as they are identical with `np.mean()` and `np.median()`, resulting in:

<details><summary>agreement_categorical()</summary>

![image](https://github.com/user-attachments/assets/f912cdec-1f85-4285-952f-4658ec5eff48)

</details>
<details><summary>agreement_numerical()</summary>

![image](https://github.com/user-attachments/assets/6cbe190c-5db6-48da-9805-2be9af71ed52)

</details>
<details><summary>evaluator_weighted_estimator()</summary>

![image](https://github.com/user-attachments/assets/3ce7fe7b-aea5-469f-adda-de9a35e46c7c)

</details>
<details><summary>mode()</summary>

![image](https://github.com/user-attachments/assets/6c3c9766-9a14-4f64-b450-6c36f1d7cecb)

</details>
<details><summary>rater_agreement_pearson()</summary>

![image](https://github.com/user-attachments/assets/7c45e575-7d4c-467d-8c8a-91e14fbf7788)

</details>

Instead of using `typing.Sequence` I used originally `numpy.typing.ArrayLike`, which is the correct one, but which results in very ugly presentations in the docs (similar to https://github.com/mitmproxy/pdoc/issues/420). As I did not found a solution, I just decided to use `typing.Sequence` for now.

It further fixes several bugs and adds more tests. There are still a lot more tests for the reliability functions needed, but as the main goal of this pull request is to get a first release of the package I will postpone this to a later stage and only focus on getting the API right here. For the review I would also propose to focus on the API and open issues for other still existing problems.

audeerington reviewed (commented):
Thanks for updating the functions, I think the input and output are a lot more practical.

The term `confidence` might be confused a bit with the confidence of a model. In past discussions, we decided to use the term `agreement` instead. But maybe it is clear enough from context what is meant here.

I also had a quick look at the documentation, and noticed that it says:

![image](https://github.com/user-attachments/assets/e758d3e8-a37d-44c3-a1a5-874538d32a07)

I agree it makes sense to focus on the API for this review, but since this is one of the first lines of the documentation, I would adjust or remove this before the release

audeerington commented on audpsychometric/core/gold_standard.py:
```diff
@@ -132,131 +123,188 @@ def evaluator_weighted_estimator(
+    ratings: typing.Sequence,
+    *,
+    axis: int = 1,
+) -> typing.Union[float, np.ndarray]:
```
```suggestion
) -> typing.Any:
```

hagenw commented on audpsychometric/core/gold_standard.py:
```diff
@@ -1,122 +1,113 @@
-
-    The returned table has an index identical to the input df,
-    and additional columns `gold_standard` and `confidence`.
+    with :math:`\text{std}` the standard deviation of the ratings.
```
Yes, good point. I wanted to write something, but I couldn't remember the correct term for it.

hagenw commented on audpsychometric/core/gold_standard.py:
```diff
@@ -1,122 +1,113 @@
-    and additional columns `gold_standard` and `confidence`.
+    The confidence for categorical data
+    is given by the fraction of raters per item
+    with the rating being equal to that of the gold standard
+    as given by :func:`audpsychometric.mode`.
```
No, `audpsychometric.mode()` will always convert to the next integer and does not support floats:
```python
>>> audpsychometric.mode([0, 1])
1
>>> audpsychometric.mode([0, 2])
1
```
As this is different from the usual mode implementation, I wanted to mention that we use our implementation here.
So maybe I change to
```suggestion
    with the rating being equal to the mode
    given by :func:`audpsychometric.mode`.
```
?

I have also updated the docstring of `mode()` to mention the rounding.

hagenw replied:
In general, our `mode()` implementation is very special, as it assumes for integer classes that they are ordered and have no missing classes, which is a strong assumption. But if met, it provides better results than the normal `mode`. I first thought about adding an argument to disable the behavior, but then a user can also just use the standard `mode` implementation of another package.

audeerington replied:
I see, thanks for clarifying! I agree with the suggestion.

I think the implementation for integer classes makes sense as it is.

The only additional feature I would like, would be some more options for the non-numerical case, when there are multiple winning categories (return a user-specified value, e.g. "no agreement", or return a random winning value rather than the first occurring one). But I was going to add an issue after this PR.

hagenw replied:
Yes, please open an issue to extend the functionality for non-numerical cases. We would need to add the argument then to `mode()` and to `agreement_categorical()` as this uses `mode()`.

hagenw commented:
I cleaned up the beginning of the API documentation by removing the discussion on dataframes:

![image](https://github.com/user-attachments/assets/b1fe6f15-8a00-4c7c-9462-b1f2481ff5dd)

I also renamed `confidence` to `agreement` in all functions.

Should I also rename `rater_agreement_pearson()` to `agreement_pearson()`, or is it better to have it separated by name to highlight that it is a slightly different concept (is it?).

audeerington commented:
>I cleaned up the beginning of the API documentation by removing the discussion on dataframes:
>
> I also renamed `confidence` to `agreement` in all functions.

Thanks!

>Should I also rename rater_agreement_pearson() to agreement_pearson(), or is it better to have it separated by name to highlight that it is a slightly different concept (is it?).

Hm, I guess I would leave it as `rater_agreement_pearson()`. I think it is a different concept since it describes how much a rater aligns with the rater consensus across multiple samples, rather than how much (dis)agreement there is for a specific sample.

ChristianGeng commented on audpsychometric/__init__.py:
```diff
@@ -2,11 +2,11 @@
 from audpsychometric.core import datasets
 from audpsychometric.core.datasets import list_datasets
 from audpsychometric.core.datasets import read_dataset
+from audpsychometric.core.gold_standard import agreement_categorical
```
The module is called `gold_standard`. 
The functions have been renamed to all start with `agreement`. 

Should the module itself as well as the test file in tests then not be renamed accordingly?
`gold_standard` was initially chosen to be in order to confirm to the table names as published using `audb.publish`, and I agree that this is not the nicest way to term it.  
But I find that the term`agreement` for me points to a kind of coefficient that should sit together with the other reliabilities.  For me what happens in this module is the calculation of a kind of `consensus` rather than agreement calculation.

hagenw replied:
Having the quote from @audeerington in mind:

> how much a rater aligns with the rater consensus across multiple samples, rather than how much (dis)agreement there is for a specific sample.

I would say, `mode()` is the only function in `gold_standard.py` that calculates a consensus at the moment, and the other functions calculate how much agreement we have for a specific sample, or how much agreement we have between raters.

ChristianGeng replied:
I will not be fussy about the exact names, but I found that there now is a kind of terminological armageddon:

```python
import audpsychometric

df = audpsychometric.read_dataset("wine")
df = df.pivot_table(index="Wine", columns="Judge", values="Scores")
print(df)
print("shape: ", df.shape)


agg = audpsychometric.agreement_categorical(df)
print("agreement_categorical", agg, agg.shape)

agg = audpsychometric.agreement_numerical(df, 0, 9)
print("agreement_numerical", agg, type(agg))

agg = audpsychometric.rater_agreement_pearson(df)
print("rater_agreement_pearson", agg, agg.shape)

agg = audpsychometric.mode(df)
print("mode value", agg, agg.shape)


agg = audpsychometric.evaluator_weighted_estimator(df)
print("ewe", agg, agg.shape)
```

This outputs
```
Judge    A    B    C    D
Wine                     
1      1.0  2.0  0.0  1.0
2      1.0  3.0  3.0  2.0
3      3.0  8.0  1.0  4.0
4      6.0  4.0  3.0  3.0
5      6.0  5.0  5.0  6.0
6      7.0  5.0  6.0  2.0
7      8.0  7.0  7.0  9.0
8      9.0  9.0  9.0  8.0
shape:  (8, 4)
agreement_categorical [0.5  0.5  0.25 0.5  0.5  0.25 0.5  0.75] (8,)
agreement_numerical 0.9037749551350623 <class 'float'>a
rater_agreement_pearson [0.85782072 0.7228383  0.83460774 0.8262462 ] (4,)
mode value [1 3 4 3 6 5 7 9] (8,)
ewe [0.96551936 2.21583295 3.85491628 4.01690183 5.51953114 5.02206086
 7.77442637 8.74510477] (8,)
```

I would have understood as a consensus the ones that have length 8 and are in the original metric. So the mean, the mode, the ewe .  I need to think a little bit more about it ( and collate some of the terms the psychologists use). This will take a little time.

hagenw replied:
Good, that you spotted this:
```
agreement_numerical 0.9037749551350623 <class 'float'>a
```
this is a bug, and should have the same size as `agreement_categorical`.

hagenw replied:
`rater_agreement_pearson` has another size as it returns a value per rater, whereas all other functions from above provide one value per stimulus.

hagenw replied:
The error with `agreement_numerical()` is surprising as we have a test for the correct returned form:

https://github.com/audeering/audpsychometric/blob/f7296ee94d522646c67c742eb4ed6cdfdf833a7c/tests/test_gold_standard.py#L124-L130

hagenw replied:
But I can confirm, that it returns the wrong shape:
```python
>>> audpsychometric.agreement_numerical([[1, 2, 3], [3, 4, 5]], 0, 10)
0.8367006838144548
```

ChristianGeng replied:
> But I can confirm, that it returns the wrong shape:
> 
> ```python
> >>> audpsychometric.agreement_numerical([[1, 2, 3], [3, 4, 5]], 0, 10)
> 0.8367006838144548
> ```

This is good in so far as it is then in the right place and should not go into reliability.

ChristianGeng reviewed (commented):
Probably caused a long time ago but I think there is a problem with the module structure:  despite doing a doing a local editable install in a fresh env, I am having an issues with running the tests. `python -m pytest` runs as expected.

However usint the pytest runner (calling `pytest`) fails:
```
ImportError while loading conftest '/home/cgeng/work/tools-public/audpsychometric/tests/conftest.py'.
tests/conftest.py:4: in <module>
    from audpsychometric import datasets
E   ModuleNotFoundError: No module named 'audpsychometric'
```

hagenw commented on audpsychometric/__init__.py:
```diff
@@ -2,11 +2,11 @@
-from audpsychometric.core.gold_standard import gold_standard_mode
-from audpsychometric.core.gold_standard import rater_confidence_pearson
+from audpsychometric.core.gold_standard import mode
+from audpsychometric.core.gold_standard import rater_agreement_pearson
```
It is not a new function, it was part of `gold_standard.py` before.
The difference to the other functions in `reliability.py` is that it does not return a single value, but one value per rater. Honestly, I don't know where to put it ;)

hagenw replied:
One solution is to remove the grouping we do for the API documentation:

![image](https://github.com/user-attachments/assets/d8b9e878-49b7-4326-ada3-e258b1a80418)

Then we don't need to agree where to put it and can just list everything alphabetically.

hagenw replied:
I think, it will be easier to discuss these in another pull request. I will merge here now and continue. Feel free to open issues for remaining questions.

ChristianGeng replied:
I think this has been opened already. It can be resolved here.
